assets/js/blocks/rating-filter/block.tsx

```diff
@@ -0,0 +1,258 @@
+	 *
+	 * @param {Array} clickedRatings Array of clicked ratings.
+	 */
+	const updateFilterUrl = ( clickedRatings: string[] ) => {
+		if ( ! window ) {
+			return;
+		}
+
+		if ( clickedRatings.length === 0 ) {
+			const url = removeQueryArgs(
+				window.location.href,
+				QUERY_PARAM_KEY
+			);
+
+			if ( url !== window.location.href ) {
+				changeUrl( url );
+			}
+
+			return;
+		}
+
+		const newUrl = addQueryArgs( window.location.href, {
+			[ QUERY_PARAM_KEY ]: clickedRatings.join( ',' ),
+		} );
+
+		if ( newUrl === window.location.href ) {
+			return;
+		}
+
+		changeUrl( newUrl );
+	};
+
+	const onSubmit = useCallback(
+		( isClicked ) => {
+			if ( isEditor ) {
+				return;
+			}
+			if ( isClicked && ! filteringForPhpTemplate ) {
+				setProductRatingsQuery( clicked );
+			}
+
+			updateFilterUrl( clicked );
+		},
+		[ isEditor, setProductRatingsQuery, clicked, filteringForPhpTemplate ]
+	);
+
+	// Track clicked STATE changes - if state changes, update the query.
+	useEffect( () => {
+		onSubmit( clicked );
+	}, [ clicked, onSubmit ] );
+
+	const clickedQuery = useMemo( () => {
+		return productRatingsQuery;
+	}, [ productRatingsQuery ] );
+
+	const currentClickedQuery = useShallowEqual( clickedQuery );
+	const previousClickedQuery = usePrevious( currentClickedQuery );
+	// Track Rating query changes so the block reflects current filters.
+	useEffect( () => {
+		if (
+			! isShallowEqual( previousClickedQuery, currentClickedQuery ) && // Clicked query changed.
+			! isShallowEqual( clicked, currentClickedQuery ) // Clicked query doesn't match the UI.
+		) {
+			setClicked( currentClickedQuery );
+		}
+	}, [ clicked, currentClickedQuery, previousClickedQuery ] );
+
+	/**
+	 * Try get the rating filter from the URL.
+	 */
+	useEffect( () => {
+		if ( ! hasSetFilterDefaultsFromUrl ) {
+			setProductRatings( initialFilters );
+			setHasSetFilterDefaultsFromUrl( true );
+		}
+	}, [
+		setProductRatings,
+		hasSetFilterDefaultsFromUrl,
+		setHasSetFilterDefaultsFromUrl,
+		initialFilters,
+	] );
+
+	/**
+	 * When a checkbox in the list changes, update state.
+	 */
+	const onClick = useCallback(
+		( clickedValue: string ) => {
+			if ( ! productRatingsArray.length ) {
+				setProductRatings( [ clickedValue ] );
+			} else {
+				const previouslyClicked =
+					productRatingsArray.includes( clickedValue );
+				const newClicked = productRatingsArray.filter(
+					( value ) => value !== clickedValue
+				);
+				if ( ! previouslyClicked ) {
+					newClicked.push( clickedValue );
+					newClicked.sort();
+				}
+				setProductRatings( newClicked );
+			}
+		},
+		[ productRatingsArray, setProductRatings ]
+	);
+
+	const heading = (
+		<TagName className="wc-block-rating-filter__title">
+			{ blockAttributes.heading }
+		</TagName>
+	);
+
+	const filterHeading = isLoading ? (
+		<FilterTitlePlaceholder>{ heading }</FilterTitlePlaceholder>
+	) : (
+		heading
+	);
+
+	let orderedRatings = [];
+	let displayedOptions = [];
```

Comment:
I'm not a TS expert so there might be a better solution. But after playing with this a bit, I think you could set the type here and then remove the `orderedRatings.filter()` below.

```suggestion
	let displayedOptions: DisplayOption[] = [];
```

You will also need to define `DisplayOption` in `rating-filter/types.ts` (and import it from this file). That `DisplayOption` type could be the same as the one in `attribute-filter/types.ts` but only with `label` and `value` properties.

Reply:
Regarding this topic, we could avoid TS errors, if we avoid mutability:

```
	const orderedRatings =
		! filteredCountsLoading &&
		objectHasProp( filteredCounts, 'rating_counts' ) &&
		Array.isArray( filteredCounts.rating_counts )
			? [ ...filteredCounts.rating_counts ].reverse()
			: [];

	const displayedOptions = orderedRatings
		.filter(
			( item ) => isObject( item ) && Object.keys( item ).length > 0
		)
		.map( ( item ) => {
			return {
				label: (
					<Rating
						className={
							productRatingsArray.includes(
								item?.rating?.toString()
							)
								? 'is-active'
								: ''
						}
						key={ item?.rating }
						rating={ item?.rating }
						ratedProductsCount={ item?.count }
					/>
				),
				value: item?.rating?.toString(),
			};
		} );
```

I suggest to created a dedicate function for doing the check regarding `orderedRatings` and give an explicit name